README.md

```diff
@@ -0,0 +1,111 @@
+  that exists for operators managed by CVO to communicate their status.
+* [openshift/library-go ClusterOperator status helpers](https://github.com/openshift/library-go/blob/master/pkg/config/clusteroperator/v1helpers/status.go)
+  that makes it easy to manage the status on a ClusterOperator resource.
+
+The goal here is to prescribe, without mandate, how to meaningfully populate the
+status of the Custom Resources your operator manages. Types, constants, and
+functions are provided for the following:
+
+* [Conditions](#Conditions)
+* [Object References](#Object-References)
+
+# Conditions
+
+Provides:
+
+* `Condition` type as specified in the [Kubernetes API Conventions](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md)
+* `ConditionType` and generally useful constants for this type (ie. "Available",
+    "Progressing", "Degraded", and "Upgradeable")
+* Functions for setting, removing, finding, and evaluating conditions.
+
+To use, simply add `Conditions` to your Custom Resource Status struct like:
+
+```
+// ExampleAppStatus defines the observed state of ExampleApp
+type ExampleAppStatus struct {
+  ...
+  // conditions describes the state of the operator's reconciliation functionality.
+  // +patchMergeKey=type
+  // +patchStrategy=merge
+  // +optional
+  // Conditions is a list of conditions related to operator reconciliation
+  Conditions []conditions.Condition `json:"conditions,omitempty"  patchStrategy:"merge" patchMergeKey:"type"`
+}
+```
+
+Then, as appropriate in your Reconcile function, use
+`conditions.SetStatusConditions` like:
+
+```
+instance := &examplev1alpha1.ExampleApp{}
+err := r.client.Get(context.TODO(), request.NamespacedName, instance)
+...handle err
+
+conditions.SetStatusCondition(&instance.Status.Conditions, conditions.Condition{
+  Type:   conditions.ConditionAvailable,
+  Status: corev1.ConditionFalse,
+  Reason: "ReconcileStarted",
+  Message: "Reconciling resource"
+})
+
+// Update the status
+err = r.client.Status().Update(context.TODO(), instance)
+...handle err
+```
+
+
+# Object References
+
+The `ObjectReference` type is provided by Kubernetes Core API
+`"k8s.io/api/core/v1"` but the functions to set and find an `ObjectReference`
+are provided. This is useful if you would like to include in the Status of your
+Custom Resource a list of objects that are managed by your operator (ie.
+Deployments, Services, other Custom Resources, etc.).
+
+For example, we can add `RelatedObjects` to our Custom Resource Status like:
+
+```
+// ExampleAppStatus defines the observed state of ExampleApp
+type ExampleAppStatus struct {
+  ...
+  // RelatedObjects is a list of objects that are "interesting" or related to this operator.
+  RelatedObjects []corev1.ObjectReference `json:"relatedObjects,omitempty"`
+}
+```
+
+Then, through Reconcile, when an object we manage has been found we can add it to
+the `RelatedObjects` slice.
```

Comment:
Should it be added when it's found?  Don't we know ahead of time?

I don't see a problem with tracking releatedObjects, but can you explain why we should track them?  Maybe you can provide an example here.

Reply:
You don't **HAVE** to have found the object ahead of time. However, it's much easier to interact with and I would contend more useful if you save the object you get back:

```
	// Check if this Subscription already exists
	found := &olm.Subscription{}
	err := r.client.Get(context.TODO(), types.NamespacedName{Name: subscription.Name, Namespace: subscription.Namespace}, found)
	if err != nil && errors.IsNotFound(err) {
	...
	}
```

Is the kind of thing you are going to have for every object that you are working with in your Reconcile function. Adding it to the relatedObjects (using the found object) looks like:

```
	// Add it to the list of RelatedObjects if found
	objectRef, err := reference.GetReference(r.scheme, found)
	if err != nil {
		return err
	}
	// Add it to the list of RelatedObjects if found
	objectreferencesv1.SetObjectReference(&instance.Status.RelatedObjects, *objectRef)
```

Semantics aside. I think that you could do it either way but I believe the idea with this library is to be opinionated about it and I like more saying, "here is the list of objects that I created and I know were created because I can get them."

Reply:
with regards to why you would track them. I think the primary purpose is the ability to audit. I know there is the idea of must-gather but, considering the HCO case, someone who doesn't know would be able to look at the HCO's CR and see all of the objects that it's managing with enough information to go inspect them.

I don't think it's a world changer, but I was grabbing the things from CVO that made sense even if not always needed. In the future, either here or in a further upstream project, I would expect there to be a lot of status things that individual operators may or may not find useful.

Reply:
I want to make sure operators get to conditions with the least amount of resistance.  Even if it's not too hard to add ObjectReferences, I think ObjectReferences should be a separate stage of integration, which I think we should be explicit about in the README.
1) Conditions - Add status to your operators CR
2) ObjectReferences - Track all the objects your operator manages.
   2a) You can use ObjectReferences for must-gather, here's how...
   2b) You can use ObjectReferences for manual object inspection, here's how...

Reply:
Ideally a great way to find stuff of course is with label selectors. An operator can put a label on all the pods is makes for example, and then it's easy for anyone to go search for pods with that label. But when using CRDs from other providers, it's less obvious to an end user what APIs the operator used, so they don't necessarily know what object types to search for. That's where I see value in tracking related objects in this way, especially when you are tracking just one instance of several different CRDs.

And of course it's optional, so in many cases I imagine users of this library may not track related objects.